src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs

```diff
@@ -1508,5 +1508,358 @@ public void nameof(string x)
+        public void TestCannotAccessStructInstancePropertyInLambdaInNameofInCSharp9()
+        {
+            var source = @"
+using System;
+
+public struct S
+{
+    public string P { get; }
+    public void M(ref string x)
+    {
+        Func<string> func = () => nameof(P.Length);
+    }
+}";
+            CreateCompilation(source, parseOptions: TestOptions.Regular9).VerifyDiagnostics(
+                // (9,42): error CS8652: The feature 'reduced member access checks in 'nameof'' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         Func<string> func = () => nameof(P.Length);
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "P").WithArguments("reduced member access checks in 'nameof'").WithLocation(9, 42));
+        }
+
+        [Fact, WorkItem(40229, "https://github.com/dotnet/roslyn/issues/40229")]
+        public void TestCannotReferenceStaticPropertyFromInstanceMemberInNameofInCSharp9()
+        {
+            var source = @"
+public class C
+{
+    public C Prop { get; }
+    public static int StaticProp { get; }
+    public string M() => nameof(Prop.StaticProp);
+}";
+            CreateCompilation(source, parseOptions: TestOptions.Regular9).VerifyDiagnostics(
+                // (6,33): error CS8652: The feature 'reduced member access checks in 'nameof'' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //     public string M() => nameof(Prop.StaticProp);
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "Prop.StaticProp").WithArguments("reduced member access checks in 'nameof'").WithLocation(6, 33));
+        }
+
+        [Fact]
+        public void TestCanReferenceStaticMethodFromInstanceMemberInNameofInCSharp9()
+        {
+            var source = @"
+public class C
+{
+    public C Prop { get; }
+    public static void StaticMethod(){}
+    public string M() => nameof(Prop.StaticMethod);
+}";
+            CreateCompilation(source, parseOptions: TestOptions.Regular9).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void TestCanAccessRefParameterInLambdaInNameof()
+        {
+            var source = @"
+using System;
+
+public struct S
+{
+    public void M(ref string x)
+    {
+        Func<string> func = () => nameof(x.Length);
+    }
+}";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(40229, "https://github.com/dotnet/roslyn/issues/40229")]
+        public void TestCanReferenceStaticMembersFromInstanceMemberInNameofUsedRecursivelyInAttributes1()
+        {
+            var source = @"
+class C
+{
+    [Attr(nameof(Prop.StaticMethod))]
+    public C Prop { get; }
+    public static void StaticMethod(){}
+}
+class Attr : System.Attribute { public Attr(string s) {} }";
+            CreateCompilation(source, parseOptions: TestOptions.RegularPreview)
+                .VerifyDiagnostics()
```

Comment:
You don't actually need to do both, `VerifyEmitDiagnostics` will get initial binding diagnostics as well.